VespaGAS.xcodeproj/project.pbxproj

```diff
@@ -8,6 +8,7 @@
 
```

Comment:
I believe this commit is interesting and I would like to discuss why.

Oil and Gasoline were not strong typed: they were just typealias of the Volume struct. 
For this simple reason, I could have changed every instance of Oil for Gasoline without any 
complain from the compiler; a first thought was to have Oil and Gasoline to inherit from a base
class Volume, but this would have introduced reference types while I wanted to work with value 
types. A better and swiftly solution has been to have a Volume protocol, a protocol extensions 
for the default implementations, and make the structs Oil and Gasoline to conform this protocol. 

While writing the protocol extensions I have noticed that several methods were returning new modified copy
of the object, and for this reason I needed to use a Generic Protocol, not just a Protocol. 

A generic protocol is not a type as a protocol or a struct, it's less type than any other type; in fact we can only use it 
as a generic constraints. In this case I haven't to face all the limitations of a Generic Protocol, instead I  only had to use the `Self` associated type when required in the methods and some little modification of the equality function. 

I was not able to reuse the equality function from volume: 
`func ==(lhs: Volume, rhs: Volume) -> Bool {
      return lhs.amount == rhs.amount && lhs.unit == rhs.unit
  }`

The compiler would have complained that now Volume has an associated type and that I could only use it as a generic constraints, the solution is to simply use it as a constraints. This concept is so new in Swift that the first time I had to think a little about it in order to understand it. To use Volume as a generic constraints is that easy as to: 
`func ==<A where A:Volume>(lhs: A, rhs: A) -> Bool {
      return lhs.amount == rhs.amount && lhs.unit == rhs.unit
  }		  
}`

Now I have most of the behaviour of Oil and Gasoline in the protocol extension of Volume, and the code I need to write to implement the new Oil and Gasoline is just few lines:
`struct Oil: Volume {
    var amount: Float
    var unit: VolumeUnit
}
`

Unfortunately I can't declare amount and unit in the protocol extension, so I have to do it into each of the types that implement the Volume protocol 😞 I don't know if there is a solution to avoid these 2 lines in each type, please tell me if you have some idea because these 2 lines are there basically to make the compiler happy.